DeepDiveLogCompiler.scala

```diff
@@ -170,29 +177,68 @@ class CompilationState( statements : DeepDiveLog.Program )  {
         ${ whereClauseStr }"""
   }
 
+  // Analyze the block visibility among statements 
+  def analyzeVisable(statements: List[Statement]) = {
+    val extractionRules   = new ListBuffer[ExtractionRule]()
+    val functionCallRules = new ListBuffer[FunctionCallRule]()
+    val inferenceRules    = new ListBuffer[InferenceRule]()
+    
+    statements foreach (_ match {
+      case s: ExtractionRule   => extractionRules += s
+      case s: FunctionCallRule => functionCallRules += s
+      case s: InferenceRule    => inferenceRules += s
+      case _                   => 
+    })
+
+    val extractionRulesGroup   = extractionRules.groupBy(_.q.head.name)
+    val functionCallRulesGroup = functionCallRules.groupBy(_.input)
+    val inferenceRulesGroup    = inferenceRules.groupBy(_.q.head.name)
+
+    extractionRulesGroup   foreach {keyVal => visable += keyVal._2(0)}
+    functionCallRulesGroup foreach {keyVal => visable += keyVal._2(0)}
+    inferenceRulesGroup    foreach {keyVal => visable += keyVal._2(0)}
+  }
```

Comment:
I think this will do:
```scala
val stmtsMap = (statements groupBy headName values) map {g => g(0) -> g} toMap
```
where `headName` is a function with cases that map each statement type to its head name.  To drop less interesting statements, We may have to filter it first based on type, or return an Option type from headName and take the values of Some keys.